Add cta banner component (#128)
primer/brand: apps/docs/content/components/CTABanner.mdx

---
title: CTABanner
status: Experimental
figma: 'https://www.figma.com/file/BJ95AjraesmRCWsKA013GS/Primer-Brand?node-id=539%3A14285&t=RJEWZhVUtrRQ0pKh-0'
source: https://github.com/primer/brand/tree/main/src/CTABanner/CTABanner.tsx

> storybook: '/brand/storybook/?path=/docs/components-ctabanner--playground'

Reviewer: This links to Docs tab, can we change this to Canvas please? We don't use the former.

```suggestion
storybook: '/brand/storybook/?path=/story/components-ctabanner--playground'
```

```

> description: Use CTABanner to create call to action content in a separated section.

Reviewer: How about this..

"Use CTABanner to draw attention to or create urgency around a user action"

cc. @ajashams @tobiasahlin @danielguillan for their suggestions also.

Author: Sounds much better to me.

---

```js
import {CTABanner, ButtonGroup, Button} from '@primer/react-brand'
```

## Examples

### Default

```jsx live
<CTABanner>
  <CTABanner.Heading>
    Where the most ambitious teams build great things
  </CTABanner.Heading>
  <CTABanner.Description>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. In sapien sit
    ullamcorper id. Aliquam luctus sed turpis felis nam pulvinar risus
    elementum.
  </CTABanner.Description>

>   <ButtonGroup buttonSize="large">

Reviewer: should `buttonSize` be configurable @danielguillan? If not, i'd recommend we set this by default inside CTABanner. Buttons feel a bit large to me, so wondering if it should be `medium` too.

Author: When I was working with Ash he said to use the large buttons medium is set by default. This is why it was created this way.

Reviewer: > When I was working with Ash he said to use the large buttons medium is set by default. This is why it was created this way.

That's fine, if large is default, can you set it _inside_ the component and remove from the docs examples? This is part of design logic, so you'd also probably want a test around that. I'd expect the only time you'd need to manually the buttonSize prop, is to change it to 'medium'.

    <Button>Primary Action</Button>
    <Button>Secondary Action</Button>
  </ButtonGroup>
</CTABanner>
```

> ### Aligned center

Reviewer: For consistency with [Hero](https://primer.style/brand/components/Hero) and other doc pages... 

```suggestion
### Alignment
```

> The content alignment can be changed using the `align` prop on the root `<ComparisonTable>`. This can be configured to either `start` or `center`.

Reviewer: ComparisonTable? 😅 

```suggestion
The content alignment can be changed using the `align` prop on the root `CTABanner`. This can be configured to either `start` or `center`.
```

```jsx live
<CTABanner align={'center'}>
  <CTABanner.Heading>
    Where the most ambitious teams build great things
  </CTABanner.Heading>
  <CTABanner.Description>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. In sapien sit
    ullamcorper id. Aliquam luctus sed turpis felis nam pulvinar risus
    elementum.
  </CTABanner.Description>
  <ButtonGroup buttonSize="large">
    <Button>Primary Action</Button>
    <Button>Secondary Action</Button>
  </ButtonGroup>
</CTABanner>
```

### Optional border

> A border can be provided using the `hasBorder` prop on the root `<ComparisonTable>`. This will render a border around the CTABanner component giving separatoin between the foreground and background. Especially when there is no shadow present on the background.

Reviewer: typo.. and reference to `ComparisonTable`

```suggestion
A border can be provided using the `hasBorder` prop on the root `<CTABanner>`. This will render a border around the CTABanner component giving further separation between the foreground and background. Especially when there is no shadow present on the background.
```

> Especially when there is no shadow present on the background.

Worth adding hasShadow={false} here to demonstrate this?

```jsx live
<CTABanner hasBorder>
  <CTABanner.Heading>
    Where the most ambitious teams build great things
  </CTABanner.Heading>
  <CTABanner.Description>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. In sapien sit
    ullamcorper id. Aliquam luctus sed turpis felis nam pulvinar risus
    elementum.
  </CTABanner.Description>
  <ButtonGroup buttonSize="large">
    <Button>Primary Action</Button>
    <Button>Secondary Action</Button>
  </ButtonGroup>
</CTABanner>
```

### Optional background

A background can be removed using the `hasBackground` prop on the root `<ComparisonTable>`. This will render the CTABanner component without a background allowing it to blend in with its parent.

```jsx live

> <CTABanner hasBackground={false}>

Reviewer: ![example lacking contrast between foreground and background](https://user-images.githubusercontent.com/13340707/204800503-ca952ade-18d5-4a06-ab43-7e3f56c96e32.png)

Not sure if this is the example I'd expect given the heading. 

Can we either provide a custom image or background color (maybe something from base scale) in the background or change the title to say "remove background fill" or similar. cc. @danielguillan for other ideas.

Also that shadow should probably be turned off here, it's a bit much 😀

Author: Oh yes so after the chat we had last week the docs were supposed to only show this with the shadow and background removed. So I will also remove the shadow. We discussed linking the two potentially if that proves to be how it is used most.

  <CTABanner.Heading>
    Where the most ambitious teams build great things
  </CTABanner.Heading>
  <CTABanner.Description>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. In sapien sit
    ullamcorper id. Aliquam luctus sed turpis felis nam pulvinar risus
    elementum.
  </CTABanner.Description>
  <ButtonGroup buttonSize="large">
    <Button>Primary Action</Button>
    <Button>Secondary Action</Button>
  </ButtonGroup>
</CTABanner>
```

### Optional shadow

> The shadow can be removed using the `hasShadow` prop on the root `<ComparisonTable>`. This will render the CTABanner component without the background shaow.

Reviewer: typo's and another reference to `ComparisonTable`

```suggestion
The shadow can be removed using the `hasShadow` prop on the root `CTABanner`. This will render the component without the background shadow.
```

```

```jsx live
<CTABanner hasShadow={false}>
  <CTABanner.Heading>
    Where the most ambitious teams build great things
  </CTABanner.Heading>
  <CTABanner.Description>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. In sapien sit
    ullamcorper id. Aliquam luctus sed turpis felis nam pulvinar risus
    elementum.
  </CTABanner.Description>
  <ButtonGroup buttonSize="large">
    <Button>Primary Action</Button>
    <Button>Secondary Action</Button>
  </ButtonGroup>
</CTABanner>
```

#### Color customization

The default colors can be customized through dedicated CSS variables.

| Name                                   | Default                          |
| :------------------------------------- | :------------------------------- |

> | `--brand-CTABanner-shadow-color-start` | `--base-color-scale-green-5-hsl` |

Author: I like default because it still tells the user what it is currently set to and is consistant with our other pages.

> | `--brand-CTABanner-shadow-color-start` | `--base-color-scale-green-5-hsl` |

Author: That makes sense I could add dark mode to it to be clear to users

> | `--brand-CTABanner-shadow-color-start` | `--base-color-scale-green-5-hsl` |

Author: These docs will be in the next push

| `--brand-CTABanner-shadow-color-end`   | `--base-color-scale-blue-5-hsl`  |

```javascript live noinline
const App = () => {
  const designTokenOverrides = `
  .custom-colors[data-color-mode='dark'] {

    /*
     * Modify the value of these tokens.
     * Remember to apply light mode equivalents if you're enabling theme switching.
     */

>     --brand-CTABanner-shadow-color-start: var(--base-color-scale-purple-5-hsl);

Reviewer: just checking that these are our official gradient pairings?

Author: I pulled it from the gradient file in Figma

    --brand-CTABanner-shadow-color-end: var(--base-color-scale-red-5-hsl);

  }
`
  return (
    <ThemeProvider colorMode="dark" className="custom-colors">
      <style>{designTokenOverrides}</style>
      <CTABanner>
        <CTABanner.Heading>
          Where the most ambitious teams build great things
        </CTABanner.Heading>
        <CTABanner.Description>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. In sapien sit
          ullamcorper id. Aliquam luctus sed turpis felis nam pulvinar risus
          elementum.
        </CTABanner.Description>
        <ButtonGroup buttonSize="large">
          <Button>Primary Action</Button>
          <Button>Secondary Action</Button>
        </ButtonGroup>
      </CTABanner>
    </ThemeProvider>
  )
}

render(<App />)
```

## Component props

> ### ComparisonTable <Label>Required</Label>

Reviewer: Heads up.. Label isn't working here. You might need to import it.
![Screenshot 2022-11-30 at 12 55 40](https://user-images.githubusercontent.com/13340707/204801913-9e1ef119-ab3e-4ac9-b231-e88ebea8c18b.png)

Reviewer: Unresolving as it's still broken.. 

https://primer-25ab31e23f-26139705.drafts.github.io/components/CTABanner#comparisontable-required

Also says ComparisonTable?

```suggestion
### CTABanner <Label>Required</Label>
```

The first row in a table is automatically styled as a table header row.

| name            | type                       | default   | required | description                                                  |
| --------------- | -------------------------- | --------- | -------- | ------------------------------------------------------------ |

> | `children`      | `ReactNode`, `ReactNode[]` | undefined | `true`   | Content to include within the banner component               |

Reviewer: syntax highlighting on undefined...

```suggestion
`| `children`      | `ReactNode`, `ReactNode[]` | `undefined` | `true`   | Content to include within the banner component               |`
```

| `align`         | `start`, `center`          | `start`   | `true`   | The layout applied to the content wihin the banner component |
| `hasBorder`     | `boolean`                  | `false`   | `false`  | A flag used to provide a border to the banner component      |
| `hasBackground` | `boolean`                  | `true`    | `false`  | A flag used to add a background to the banner component      |
| `className`     | `string`                   |           | `false`  | CTABanner custom class                                       |

> ### ComparisonTable.Heading

Reviewer: ```suggestion
### CTABanner.Heading
```

Optional node that can be used to provide a heading for the banner.

| name        | type                       | default   | required | description                                                           |
| ----------- | -------------------------- | --------- | -------- | --------------------------------------------------------------------- |
| `className` | `string`                   |           | `false`  | Custom class name for the ComparisonTable.Heading component           |
| `children`  | `ReactNode`, `ReactNode[]` | undefined | `true`   | Content to be displayed inside the ComparisonTable.Heading component. |

> | `as`        | `h1`, `h2`                 | `h1`      | `false`  | Content to be displayed inside the column cell.                       |

Reviewer: related feedback https://github.com/primer/brand/pull/128/files#r1035940710

Author: Other than the default which we will discuss in maintainers sync I updated it to include all options.

```javascript live noinline
const App = () => {
  const designTokenOverrides = `
  .custom-colors[data-color-mode='dark'] {

    /*
     * Modify the value of these tokens.
     * Remember to apply light mode equivalents if you're enabling theme switching.
     */
    --brand-CTABanner-shadow-color-start: var(--base-color-scale-purple-5-hsl);
    --brand-CTABanner-shadow-color-end: var(--base-color-scale-red-5-hsl);

  }
`
  return (
    <ThemeProvider colorMode="dark" className="custom-colors">
      <style>{designTokenOverrides}</style>
      <CTABanner>
        <CTABanner.Heading>
          Where the most ambitious teams build great things
        </CTABanner.Heading>
        <CTABanner.Description>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. In sapien sit
          ullamcorper id. Aliquam luctus sed turpis felis nam pulvinar risus
          elementum.
        </CTABanner.Description>
        <CTABanner.ButtonGroup>
          <Button>Primary Action</Button>
          <Button>Secondary Action</Button>
        </CTABanner.ButtonGroup>
      </CTABanner>
    </ThemeProvider>
  )
}

render(<App />)
```

## Component props

> ### CTABanner <Label>Required</Label>

Reviewer: My previous comments about this seem to have disappeared 😕 

This is breaking in the rendered version because Label isn't imported.

fix is - from memory (try it locally) - to add `import {Label} from "@primer/react"` at the top of the file. 

![Screenshot 2022-12-07 at 13 38 53](https://user-images.githubusercontent.com/13340707/206194145-7a943e61-2002-44c2-88f8-b240c76f5949.png)

Author: Ahh I thought I fixed this but I imported it from primer-brand which didn't make sense.

Author: This will be in the next push

---
title: CTABanner
status: Experimental
figma: 'https://www.figma.com/file/BJ95AjraesmRCWsKA013GS/Primer-Brand?node-id=539%3A14285&t=RJEWZhVUtrRQ0pKh-0'
source: https://github.com/primer/brand/tree/main/src/CTABanner/CTABanner.tsx
storybook: '/brand/storybook/?path=/story/components-ctabanner--playground'
description: Use CTABanner to draw attention to or create urgency around a user action.
---

```js
import {CTABanner, Button} from '@primer/react-brand'

> import {Label} from '@primer/react'

Reviewer: This needs to be outside the code snippet scope.. it's currently rendering in the docs..

![Screenshot 2022-12-07 at 15 12 21](https://user-images.githubusercontent.com/13340707/206216563-4317fbaa-4531-40f6-b95b-f5045507f48b.png)
